yabox/algorithms/de.py

```diff
@@ -164,54 +117,95 @@ def evaluate_denormalized(self, PD):
-        else:
-            iterator = self.iterator()
-        for step in iterator:
-            idx = step.best_idx
-            P = step.population
-            fitness = step.fitness
-            if step.iteration > self.maxiters:
-                if show_progress:
-                    iterator.n = self.maxiters
-                    iterator.refresh()
-                    iterator.close()
-                return self.denormalize(P[idx].reshape(-1, 1)), fitness[idx]
+            iterator = tqdm(self.iterator(), total=self.maxiters,
+                            desc=f'Optimizing ({self.name})')
+
+        # Check https://docs.python.org/3/library/itertools.html#itertools-recipes
+        # feed the entire iterator into a zero-length deque
+        deque(iterator, maxlen=0)
+              
+        if show_progress:
+            iterator.n = self.maxiters
+            iterator.refresh()
+            iterator.close()
+        
+        best_ind = self.population[self.best_idx].reshape(-1, 1)
+        best_ind_denorm = self.denormalize(best_ind)
+        return best_ind_denorm, self.best_fitness
 
 
 class PDE(DE):
     def __init__(self, fobj, bounds, mutation=(0.5, 1.0), crossover=0.7, maxiters=1000,
                  self_adaptive=False, popsize=None, seed=None, processes=None, chunksize=None):
-        super().__init__(fobj, bounds, mutation, crossover, maxiters, self_adaptive, popsize, seed)
+        
         from multiprocessing import Pool
         self.processes = processes
         self.chunksize = chunksize
         self.name = 'Parallel DE'
         self.pool = None
         if processes is None or processes > 0:
             self.pool = Pool(processes=self.processes)
+        
+        super().__init__(fobj, bounds, mutation, crossover, maxiters, self_adaptive, popsize, seed)
+
+        self.mutants = np.zeros((self.popsize, self.dims))
+        
+    
+    def create_mutant(self, i):
+        mutant = super().create_mutant(i)
+        # Add to the mutants population for parallel evaluation (later)
+        # self.mutants.append(mutant)
+        self.mutants[i, :] = mutant
+
+    def replace(self):
+        # Evaluate the whole new population (class PDE implements a parallel version of evaluate)
+        mutant_fitness = self.evaluate(self.mutants)
+        
+        # We have to make super().replacement static in other to parallelized it
+        # self.pool.starmap(super().replacement, zip(range(self.popsize), self.mutants, mutant_fitness), chunksize=self.chunksize)
+        for j, (mutant, fitness) in enumerate(zip(self.mutants, mutant_fitness)):
+            super().replacement(j, mutant, fitness)
 
     def iterator(self):
-        it = PDEIterator(self)
-        try:
-            for data in it:
-                yield data
-        finally:
-            if self.pool is not None:
-                self.pool.terminate()
+        # This is the main DE loop. For each vector (target vector) in the
```

Comment:
This is duplicated code. I've used DEIterator and PDEIterator to avoid this kind of duplications.

Reply:
The duplicate code is minimal, just a few lines, but it is much more readable and expressive than to use two more classes. Explicit is better than implicit, and in the other solutions there are some things I found kind of forced, like asking everytime in the replace method of PDE if it's the last element just to avoid re writing the while loop. If you have a population of n elements you are adding n comparisons per iteration and could leave to performance issues.

Reply:
Also remember thenzen of python (just a fragment):

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex
Flat is better than nested.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.

That was my main motivation for the changes especially: Explicit is better than implicit and Readability counts

Reply:
I forgot to mention that the main idea of using those clases was not only to avoid code duplication but to also to separate the concept of the algorithm (config) from the iterator. In this way, you can pass the same DE() instance to different methods that can reuse it to create iterators without the need of using an algorithm builder.

Note that I've also replaced the line:
`yield self.population, self.fitness, self.best_idx`
with
`yield self`

In this way, it's easier to access the algorithm internal state (I'm using this approach now to collect information during the execution of the algorithm). You can still do that but at expenses of mixing the algorithm config with the internal state. I also agree that checking all the time if the vector is the last one in the replace function is kind of forced, but in your approach both the semantics and syntax of the replace method are different (in DE and PDE) and I find it a little bit confusing. For me this is a sign that there is another way to design it.

I don't think your option is bad, it very interesting and well done. I like most of the changes you did, but I'm still not 100% convinced about removing the DEIterator and PDEIterator. I will think about it

Reply:
I get your point, so it's true that separating the internal state from the config is a good idea but I don't think having two extra classes is the solution, I will think about another way of doing it.

Another key part is understand what parallel programming implies, OOP isn't usually quite compatible with parallel programming, for parallel execution it is better to have a functional programming approach and that's why PDE would not be "very similar" with DE neither in the algorithm nor in the syntax. Because in DE one should think "Vector by vector" and in PDE it's rather "All vectors at once". This all at once approach is commonly used in numpy but numpy isn't enough for this case and we should you strategies as pool.map and pool.starmap and that implies that we can't be pure OOP.

So in PDE the steps would be something like:

- Generate all the trial vectors at once
- Crossover all vectors at once
- Calculate all fitness at once
- Replace them when necessary all at once
- Extract the best idx and the best fitness

So we can't have the notion of "iteration of each vector" (the for loop)  because we can't control in which order the computations would be completed, it would be much faster with several CPUs and with a complex situation (expensive fobj or lots of population).

In this code only the evaluate is parallelized but I'm working in a parallelized version of create_mutant and replace.

Reply:
Thank you so much for your interesting review! Just one thing: the main idea of the PDE is to parallelize the evaluation of the vectors, not all the operations, since the evaluation is (in most cases) the main bottleneck. Actually, in the research literature this is usually known as "master-slave" parallelization. There is no much interest on parallelizing every operation. Instead, what people usually do is to run differenet populations in a distributed way, sharing the best solutions between populations (the island model), see for example http://esa.github.io/pagmo

Reply:
Excelent, thanks for the example and the key concepts, I believe it's a good idea to merge the first 3 commits and I will do the following:

- Think of another way of separating the internal state from the config, keeping OOP
- Try to parallelize with numpy as much as possible
- Try with a FP approach, I think that classes are not always the best for these scenarios and maybe, MAYBE some module level functions with some dictionaries could do the trick, also this way it would be much easier to parallelize and to implement things like async/await (in cases where evaluation implies long but non CPU intensive tasks)

I will add some more branches and work on my fork and submit PR as I finish the features